### ISSUE TITLE
AttributeError: 'builtins.Certificate' object has no attribute '_x509'

### ISSUE THREAD

**Issue by abmuslim** [aiortc/aiortc#806]
I am working on the WebRTC application. It is working fine on other Linux versions. But when I installed it in OpenSuse, it give me this error related to certificates.

### **Error**

INFO:aiohttp.access:185.134.139.187 [03/Jan/2023:10:14:09 +0000] "GET / HTTP/1.1" 200 2304 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
INFO:aiohttp.access:185.134.139.187 [03/Jan/2023:10:14:09 +0000] "GET /client.js HTTP/1.1" 200 6577 "http://193.25.X.X/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
ERROR:aiohttp.server:Error handling request
Traceback (most recent call last):
  File "/home/sysadmin/.local/lib/python3.10/site-packages/aiohttp/web_protocol.py", line 433, in _handle_request
    resp = await request_handler(request)
  File "/home/sysadmin/.local/lib/python3.10/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
  File "/home/sysadmin/abdullah/webrtc-server/src/server.py", line 105, in return_offer
    pc.addTransceiver(trackOrKind="video")
  File "/home/sysadmin/.local/lib/python3.10/site-packages/aiortc/rtcpeerconnection.py", line 471, in addTransceiver
    return self.__createTransceiver(
  File "/home/sysadmin/.local/lib/python3.10/site-packages/aiortc/rtcpeerconnection.py", line 1067, in __createTransceiver
    dtlsTransport = self.__createDtlsTransport()
  File "/home/sysadmin/.local/lib/python3.10/site-packages/aiortc/rtcpeerconnection.py", line 1044, in __createDtlsTransport
    dtlsTransport = RTCDtlsTransport(iceTransport, self.__certificates)
  File "/home/sysadmin/.local/lib/python3.10/site-packages/aiortc/rtcdtlstransport.py", line 375, in __init__
    self.__ctx = certificate._create_ssl_context()
  File "/home/sysadmin/.local/lib/python3.10/site-packages/aiortc/rtcdtlstransport.py", line 211, in _create_ssl_context
    _openssl_assert(lib.SSL_CTX_use_certificate(ctx, self._cert._x509) == 1)  # type: ignore
**AttributeError: 'builtins.Certificate' object has no attribute '_x509'**
INFO:aiohttp.access:185.134.139.187 [03/Jan/2023:10:14:14 +0000] "GET /data/CLIENT HTTP/1.1" 500 245 "http://193.25.X.X/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"

### **Versions**
**openssl =** OpenSSL 1.1.1s  1 Nov 2022
**OS =** OpenSuse-Linux. 
**python3 -c 'import OpenSSL; print(OpenSSL.__version__)'** = 23.0.0

**Comment by abmuslim** [comment]
The issue is solved with the help of https://github.com/aiortc/aiortc/issues/804

**Comment by chschnell** [comment]
As this is my first post here I'd first like to thank the founder of this project and all contributers for this excellent and highly useful library (that also includes aioice).

I think there is a proper way to fix this using [pyOpenSSL](https://www.pyopenssl.org/en/latest/).

In cryptography version 39.0.0 private member `_x509` was dropped from class cryptography.x509.Certificate, it kept a reference to OpenSSL's low-level C struct of type `<cdata 'X509 *'>`.

pyOpenSSL offers methods to convert [cryptography.x509.Certificate](https://cryptography.io/en/latest/x509/reference/#x-509-certificate-object) objects to and from [OpenSSL.crypto.X509](https://www.pyopenssl.org/en/latest/api/crypto.html#x509-objects) objects:

* [X509.from_cryptography(Certificate) → X509](https://www.pyopenssl.org/en/latest/api/crypto.html#OpenSSL.crypto.X509.from_cryptography)
* [X509.to_cryptography() → Certificate](https://www.pyopenssl.org/en/latest/api/crypto.html#OpenSSL.crypto.X509.to_cryptography)

And OpenSSL.crypto.X509 objects keep a reference to that low-level `<cdata 'X509 *'>` C struct of OpenSSL (the one that went missing) under the same private name that cryptography.x509.Certificate used to, `_x509`.

Steps to make rtcdtlstransport.py compatible with cryptography 39.0.0:

1. install pyopenssl v23 and cryptography v39 using
    `pip uninstall -y pyopenssl cryptography`
    `pip install pyopenssl cryptography`
2. add pyOpenSSL import statement below [line 19](https://github.com/aiortc/aiortc/blob/f4e3049875142a18fe32ad5f2c052b84a3112e30/src/aiortc/rtcdtlstransport.py#L19):
    `import OpenSSL.crypto`
3. below [line 169](https://github.com/aiortc/aiortc/blob/f4e3049875142a18fe32ad5f2c052b84a3112e30/src/aiortc/rtcdtlstransport.py#L169) add new member:
    `self._openssl_cert = OpenSSL.crypto.X509.from_cryptography(cert)`
4. change [line 186](https://github.com/aiortc/aiortc/blob/f4e3049875142a18fe32ad5f2c052b84a3112e30/src/aiortc/rtcdtlstransport.py#L186) from:
    `value=certificate_digest(self._cert._x509),`
   to:
    `value=certificate_digest(self._openssl_cert._x509),`
5. change [line 211](https://github.com/aiortc/aiortc/blob/f4e3049875142a18fe32ad5f2c052b84a3112e30/src/aiortc/rtcdtlstransport.py#L211) from:
    `_openssl_assert(lib.SSL_CTX_use_certificate(ctx, self._cert._x509) == 1)`
   to:
    `_openssl_assert(lib.SSL_CTX_use_certificate(ctx, self._openssl_cert._x509) == 1)`

I've tested this on a Raspberry Pi with webcam.py (how should I build aiortc using setup.py under windows, where should I install the opus and vpx library sources?), python packages used:

    $ pip show pyopenssl cryptography
    Name: pyOpenSSL
    Version: 23.0.0
    Summary: Python wrapper module around the OpenSSL library
    Home-page: https://pyopenssl.org/
    Author: The pyOpenSSL developers
    Author-email: cryptography-dev@python.org
    License: Apache License, Version 2.0
    Location: /home/sdbian/.local/lib/python3.9/site-packages
    Requires: cryptography
    Required-by:
    ---
    Name: cryptography
    Version: 39.0.0
    Summary: cryptography is a package which provides cryptographic recipes and primitives to Python developers.
    Home-page: https://github.com/pyca/cryptography
    Author: The Python Cryptographic Authority and individual contributors
    Author-email: cryptography-dev@python.org
    License: (Apache-2.0 OR BSD-3-Clause) AND PSF-2.0
    Location: /home/sdbian/.local/lib/python3.9/site-packages
    Requires: cffi
    Required-by: aiortc, pyOpenSSL

Before making a PR for this I'd like to share this idea here for discussion.

Actually I believe that for a better solution rtcdtlstransport.py should be rebased from the low-level OpenSSL C bindings to pyOpenSSL.

Most of the low-level `lib.XY()` functions used in rtcdtlstransport.py map directly to methods of pyOpenSSL's classes [OpenSSL.SSL.Context](https://www.pyopenssl.org/en/latest/api/ssl.html#context-objects), [OpenSSL.SSL.Connection](https://www.pyopenssl.org/en/latest/api/ssl.html#connection-objects) and [OpenSSL.crypto.X509](https://www.pyopenssl.org/en/latest/api/crypto.html#x509-objects), I think only [DTLSv1_get_timeout()](https://www.openssl.org/docs/manmaster/man3/DTLSv1_get_timeout.html) and [DTLSv1_handle_timeout()](https://www.openssl.org/docs/manmaster/man3/DTLSv1_handle_timeout.html) are missing and would need to be kept as low-level C bindings (this needs further investigation, but I have the impression pyOpneSSL doesn't use these two at all).

Reading and writing to and from the SSL connection would also have to be redesigned and simplified, but BIO connections are supported so it looks good (from the distance).

This would simplify the SSL code in rtcdtlstransport.py (pyOpenSSL brings all the boilerplate code, memory management and error Exceptions) and should make it more robust for future changes.

So far I have only sketched this idea out in my text editor, before giving this a real shot I would even more like to put this idea out here for discussion. Maybe someone else went down this road before and there are problems ahead that I don't see yet?

**Comment by jlaine** [comment]
This is a duplicate of #804
